### PR TITLE
Hardcode painless as the default scripting lang and add legacy script default for stored scripts

### PR DESCRIPTION
**Pull request by martijnvg**
- Removed `script.default_lang` setting and made `painless` the hardcoded default script language.
  - The default script language is now maintained in `Script` class.
- Added `script.legacy.default_lang` setting that controls the default language for scripts that are stored inside documents (for example percolator queries).  This defaults to groovy.
  - Added `QueryParseContext#getDefaultScriptLanguage()` that manages the default scripting language. Returns always `painless`, unless loading query/search request in legacy mode then the returns what is configured in `script.legacy.default_lang` setting.
  - In the aggregation parsing code added `ParserContext` that also holds the default scripting language like `QueryParseContext`. Most parser don't have access to `QueryParseContext`. This is for scripts in aggregations.
- The `lang` script field is always serialized (toXContent).

PR for #20122

Need to add tests and docs, but before adding those I'd like to get some feedback.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/script/ScriptSettings.java [hunk 1]
```diff
@@ -160,15 +160,19 @@ public ScriptSettings(ScriptEngineRegistry scriptEngineRegistry, ScriptContextRe
...
     }
 
-    public Setting<String> getDefaultScriptLanguageSetting() {
-        return defaultScriptLanguageSetting;
+    public Setting<String> getDefaultLegacyScriptLanguageSetting() {
+        return defaultLegacyScriptLanguageSetting;
+    }
+
+    public static String getLegacyDefaultLang(Settings settings) {
+        return settings.get("script.legacy.default_lang", ScriptSettings.LEGACY_DEFAULT_LANG);
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/script/ScriptSettings.java]: this key is different from the one specified above in https://github.com/elastic/elasticsearch/pull/20310/files#diff-9a1a4781a78e7e7a89657e8717f88997R61

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/percentiles/AbstractPercentilesParser.java [hunk 2]
```diff
@@ -45,10 +45,11 @@ public AbstractPercentilesParser(boolean formattable) {
...
 
     @Override
-    protected boolean token(String aggregationName, String currentFieldName, Token token, XContentParser parser,
-            ParseFieldMatcher parseFieldMatcher, Map<ParseField, Object> otherOptions) throws IOException {
+    protected boolean token(String aggregationName, String currentFieldName, Token token,
+                            ParserContext context, Map<ParseField, Object> otherOptions) throws IOException {
+        XContentParser parser = context.getParser();
         if (token == XContentParser.Token.START_ARRAY) {
-            if (parseFieldMatcher.match(currentFieldName, keysField())) {
+            if (context.getParseFieldMatcher().match(currentFieldName, keysField())) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/percentiles/AbstractPercentilesParser.java]: I wonder if we should have a method in ParserContext#matchField(...) that calls `getParseFieldMatcher().match` for us?

File: modules/percolator/src/main/java/org/elasticsearch/percolator/PercolatorFieldMapper.java [hunk 3]
```diff
@@ -270,7 +270,9 @@ public Mapper parse(ParseContext context) throws IOException {
         }
 
         XContentParser parser = context.parser();
-        QueryBuilder queryBuilder = parseQueryBuilder(queryShardContext.newParseContext(parser), parser.getTokenLocation());
+        QueryBuilder queryBuilder = parseQueryBuilder(
+                queryShardContext.newParseContext(parser), parser.getTokenLocation()
```
Comment (s1monw) [inline modules/percolator/src/main/java/org/elasticsearch/percolator/PercolatorFieldMapper.java]: we still need to somehow parse the incoming queries and rewrite their source before we store them no? I mean we can do this in a different PR but we somehow need to parse Percolator queries differently if they are coming from an API or from an index. If they come from an API we need to parse with painless as default and then rewrite the source to be explicit?

File: core/src/main/java/org/elasticsearch/search/aggregations/support/ParserContext.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.search.aggregations.support;
+
+import org.elasticsearch.common.ParseFieldMatcher;
+import org.elasticsearch.common.xcontent.XContentParser;
+
+public final class ParserContext {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/search/aggregations/support/ParserContext.java]: maybe call it XContentParseContext

File: modules/percolator/src/main/java/org/elasticsearch/percolator/PercolatorFieldMapper.java (hunk 3 above)
Comment (martijnvg) [inline modules/percolator/src/main/java/org/elasticsearch/percolator/PercolatorFieldMapper.java]: > we still need to somehow parse the incoming queries and rewrite their source before we store them no?

Luckily we already rewrite the source already at line 281

> If they come from an API we need to parse with painless as default and then rewrite the source to be explicit?

Because of the rewrite at line 281 and the fact the `lang` param is explicit set to `painless` if no `lang` has been specified in the Script class, all percolator queries from 5.0 and up will have explicit lang in any `script` query or `script` score function.

Percolator queries stored in indices created prior the ES 5 will need to be reindexed at some point after the upgrade. These percolator queries don't use the `percolator` field type and use the `object` field type, so we can't rewrite those queries when updated. However when these queries are used by the `percolate` query we pick the legacy script language (PercolatorQueryBuilder line 518). 

I think this approach is ok as we already strongly advice to reindex any 2.x percolator index in the reference documentation and migrate docs?

File: core/src/main/java/org/elasticsearch/script/ScriptSettings.java (hunk 1 above)
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/script/ScriptSettings.java]: oops, will fix

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/percentiles/AbstractPercentilesParser.java (hunk 2 above)
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/percentiles/AbstractPercentilesParser.java]: +1

File: core/src/main/java/org/elasticsearch/search/aggregations/support/ParserContext.java (hunk 4 above)
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/search/aggregations/support/ParserContext.java]: +1

File: modules/percolator/src/main/java/org/elasticsearch/percolator/PercolatorFieldMapper.java (hunk 3 above)
Comment (s1monw) [inline modules/percolator/src/main/java/org/elasticsearch/percolator/PercolatorFieldMapper.java]: you are the man! that is awesome!!! that should just work. I really wonder if we can build a BWC test index with a percolator that ensures we can read this stuff if would be awesome to have asuch a test

File: core/src/main/java/org/elasticsearch/script/ScriptSettings.java [hunk 5]
```diff
@@ -32,7 +32,8 @@
 
 public class ScriptSettings {
 
-    public static final String LEGACY_DEFAULT_LANG = "groovy";
+    static final String LEGACY_DEFAULT_LANG = "groovy";
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/script/ScriptSettings.java]: can you please put a big comment on this that this language is ONLY used for scripts stored in the percolator index etc. not for scripts coming in via the API

File: core/src/main/java/org/elasticsearch/search/aggregations/support/XContentParseContext.java [hunk 6]
```diff
@@ -47,4 +52,9 @@ public ParseFieldMatcher getParseFieldMatcher() {
     public String getDefaultScriptLanguage() {
         return defaultScriptLanguage;
     }
+
+    public boolean matchField(String fieldName, ParseField parseField) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/search/aggregations/support/XContentParseContext.java]: please add some javadocs

**Comment by s1monw** [comment]
I really think we need a test for this. maybe we can build a bwc index with a script in the percolator? Maybe we can even fake it in a unittest and build a lucene index ourself, start an indexservice etc. to test this. Otherwise this looks awesome

File: modules/percolator/src/main/java/org/elasticsearch/percolator/PercolatorFieldMapper.java (hunk 3 above)
Comment (martijnvg) [inline modules/percolator/src/main/java/org/elasticsearch/percolator/PercolatorFieldMapper.java]: The `PercolatorBackwardsCompatibilityTests` already exists for verifying that pre 5 indices remain to work. My plan is to adjust this test, so that we also check we do the right things for scripts in percolator query with implicit default lang.

**Comment by s1monw** [comment]
I think we can remove the WIP label here too

**Comment by martijnvg** [comment]
@s1monw I've updated the PR and replaced the WIP label with review label.

**Comment by s1monw** [comment]
LGTM great solution, good tests, well done! thanks mate!
